# Laboratorio de regresión logística

|                |   |
:----------------|---|
| **Nombre**     | Michelle Trigo  |
| **Fecha**      | 08/10/2025  |
| **Expediente** | 745567  | 

In machine learning, Support Vector Machines (SVM) are supervised learning models with associated learning algorithms that analyze data used for classification and regression analysis. It is mostly used in classification problems. In this algorithm, each data item is plotted as a point in p-dimensional space (where p is the number of features), with the value of each feature being the value of a particular coordinate. Then, classification is performed by finding the hyper-plane that best differentiates the two classes (or more if we have a multi class problem):

$$ f(x) = w^T \varphi(x) + b $$

where $\varphi: X \rightarrow F $ is a function that makes each input point $x$ correspond to a point in F, where F is a Hilbert space.

In addition to performing linear classification, SVMs can efficiently perform a non-linear classification, implicitly mapping their inputs into high-dimensional feature spaces (more specifically using the kernel trick, like the RBF funcion). 

[1]

OLS utilizes the squared residuals to fit the parameters. Large residuals caused by outliers may worsen the accuracy significantly.

Support Vectors use piecewise linear functions to counter this, in which a hyperparameter  $\epsilon$ called the margin lets errors that are less or equal to it be 0, and error larger than it be $e - \epsilon$. 

The problem to solve is:

\begin{split}
        \min_{w, b, \xi, \xi^*} \mathcal{P}_\epsilon(w, b, \xi) &= \frac{1}{2} w^T w + c \sum_{k=1}^{N} \xi_k \\
        \text{s.t. } & y_k [w^T \varphi(x_k) - b] \geq 1- \xi_k,\ \ k = 1, ..., N \\
        & \xi_k \geq 0,\ \ k = 1, ..., N
\end{split}


The most important question that arises when using a SVM is how to choose the correct hyperplane. Consider the following scenarios:

### Scenario 1

In this scenario there are three hyperplanes called A, B, and C. Now, the problem is to identify the hyperplane which best differentiates the stars and the circles.

<center><img src="https://media.geeksforgeeks.org/wp-content/uploads/SVM_21-2.png" alt="what image shows"></center>

In this case, hyperplane B separates the stars and the circle betters, hence it is the correct hyperplane.


### Scenario 2

Now take another scenario where all three hyperplanes are segregating classes well. The question that arises is how to choose the best hyperplane in this situation.

<center><img src="https://media.geeksforgeeks.org/wp-content/uploads/SVM_4-2.png" alt="what image shows"></center>

In such scenarios, we calculate the margin (which is the distance between nearest data point and the hyperplane). The hyperplane with the largest margin will be considered as the correct hyperplane to classify the dataset.

Here C has the largest margin. Hence, it is considered as the best hyperplane.


### Kernels
Knowing 
$$ w = \sum_{k=1}^{N} \alpha_k y_k \varphi(x_k) $$

And
$$ y_{pred} = w^T \varphi(x) + b $$

Then 
$$ y_{pred} = (\sum_{k=1}^{N} \alpha_k y_k \varphi(x_k))^T \varphi(x) + b $$

Where $\varphi$ is a function that makes each input in $x$ correspond to a point in $F$ (a Hilbert space). This can be seen as processing and transforming the input featuers to keep the model's convexity. [2]

This also allows us to transform the inputs into another space where they might be more easily classified.

<center><img src=https://miro.medium.com/max/838/1*gXvhD4IomaC9Jb37tzDUVg.png alt="what image shows"></center>

## ROC and AUC

A ROC (Receiver Operating Characteristic) is a graph that shows how the classification model performs at the classification thresholds. 

ROC curves typically feature true positive rate on the Y axis, and false positive rate on the X axis. This means that the top left corner of the plot is the “ideal” point - a false positive rate of zero, and a true positive rate of one. This is not very realistic, but it does mean that a larger area under the curve (AUC) is usually better. [3]

True Positive Rate is a synonym for Recall and defined as:
$$ TPR = \frac{TP}{TP + FN} $$

False Positive Rate is a synonym for Specificity and defined as:

$$ FPR = \frac{FP}{FP + TN} $$

ROC curves are typically used in binary classification to study the output of a classifier. In order to extend ROC curve and ROC area to multi-label classification, it is necessary to binarize the output. One ROC curve can be drawn per label, but one can also draw a ROC curve by considering each element of the label indicator matrix as a binary prediction (micro-averaging).

E.g. If you lower a classification threshold, more items would be classified as positive, increasing False Positives and True Positives.

AUC stands for Area under the ROC.

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
import statsmodels.api as sm
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import Ridge, Lasso
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.model_selection import cross_val_score

## Ejercicio 1

- Utiliza el dataset `Iris`, modela con SVC y haz Cross-Validation de diferentes kernels ('linear', 'poly', 'rbf', 'sigmoid').
- Modela con LogisticRegression.
- El método de Cross-Validation es K-Folds con $k=10$.
- Utiliza el AUC como métrico de Cross-Validation.
- Compara resultados.

In [4]:
from sklearn.svm import SVC
from sklearn.datasets import load_iris 
from sklearn.model_selection import cross_val_score

In [4]:
load_iris?

Utiliza el dataset Iris, modela con SVC y haz Cross-Validation de diferentes kernels ('linear', 'poly', 'rbf', 'sigmoid').

In [28]:
iris = load_iris()
X = iris.data
y = iris.target

kernels = ['linear', 'poly', 'rbf', 'sigmoid']

for k in kernels:
    model = SVC(kernel=k, degree=2)  
    scores = cross_val_score(model, X, y, cv=5)  # 5-fold cross-validation
    print(f"Kernel: {k:7s} | Accuracy promedio: {scores.mean():.4f} | Desviación: {scores.std():.4f}")

Kernel: linear  | Accuracy promedio: 0.9800 | Desviación: 0.0163
Kernel: poly    | Accuracy promedio: 0.9867 | Desviación: 0.0163
Kernel: rbf     | Accuracy promedio: 0.9667 | Desviación: 0.0211
Kernel: sigmoid | Accuracy promedio: 0.0667 | Desviación: 0.0596


Modela con LogisticRegression.


In [18]:
model = LogisticRegression(max_iter=200)
scores = cross_val_score(model, X, y, cv=5)

print("Accuracy promedio:", scores.mean())

Accuracy promedio: 0.9733333333333334


El método de Cross-Validation es K-Folds con $k=10$.


In [11]:
from sklearn.model_selection import KFold, cross_val_score

In [25]:
kf = KFold(n_splits=10, shuffle=True, random_state=42)

model = LogisticRegression(max_iter=200)

scores = cross_val_score(model, X, y, cv=kf)

print("Accuracy promedio:",scores.mean())
print("Desviación estándar:",scores.std())

Accuracy promedio: 0.9733333333333334
Desviación estándar: 0.04422166387140532


Utiliza el AUC como métrico de Cross-Validation.


In [26]:
scores = cross_val_score(
    model, X, y, cv=kf, scoring='roc_auc_ovr'
)

print("AUC promedio:",scores.mean())
print("Desviación estándar: ",scores.std())

AUC promedio: 0.9974126182459516
Desviación estándar:  0.005187479633482653


Compara resultados

Tanto los SVM como la regresión logística clasifican muy bien el dataset, con accuracies alrededor de 0.97–0.99; el kernel polinómico del SVM fue el más preciso, seguido del lineal y RBF, mientras que el kernel sigmoid no funciona. 

La regresión logística con K-Folds 10 también mostró un desempeño excelente y estable, y al medirlo con AUC alcanzó casi 1.0, lo que indica que separa casi perfectamente las tres clases de flores.


## Ejercicio 2
- Repite el ejercicio 1 con el dataset `Default`. Utiliza `default` como target.

In [5]:
df = pd.read_csv("Default.csv")
df.head()

default student      balance        income
0      No      No   729.526495  44361.625074
1      No     Yes   817.180407  12106.134700
2      No      No  1073.549164  31767.138950
3      No      No   529.250605  35704.493940
4      No      No   785.655883  38463.495880

Haz Cross-Validation de diferentes kernels ('linear', 'poly', 'rbf', 'sigmoid').

In [6]:
# Convertir variables categóricas a numéricas
df['default'] = df['default'].map({'No':0,'Yes':1})
df['student'] = df['student'].map({'No':0,'Yes':1})

# Separar X y y
X = df.drop(columns=['default'])
y = df['default']

# Escalar
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

In [7]:
kernels = ['linear', 'poly', 'rbf', 'sigmoid']

for k in kernels:
    model = SVC(kernel=k, degree=2)  # degree solo aplica al 'poly'
    scores = cross_val_score(model, X_scaled, y, cv=5)  # 5-fold por el tamaño del dataset
    print(f"Kernel: {k:7s} | Accuracy promedio: {scores.mean():.4f} | Desviación: {scores.std():.4f}")

Kernel: linear  | Accuracy promedio: 0.9667 | Desviación: 0.0002
Kernel: poly    | Accuracy promedio: 0.9694 | Desviación: 0.0010
Kernel: rbf     | Accuracy promedio: 0.9716 | Desviación: 0.0017
Kernel: sigmoid | Accuracy promedio: 0.9536 | Desviación: 0.0033


Modela con LogisticRegression.

In [8]:
model = LogisticRegression(max_iter=200)
scores = cross_val_score(model, X_scaled, y, cv=5)

print("Accuracy promedio:", scores.mean())
print("Desviación:", scores.std())

Accuracy promedio: 0.9731
Desviación: 0.0014282856857085744


El método de Cross-Validation es K-Folds con  𝑘=10

In [12]:
kf = KFold(n_splits=10, shuffle=True, random_state=42)

model = LogisticRegression(max_iter=200)

scores = cross_val_score(model, X_scaled, y, cv=kf)

print("Accuracy promedio:",scores.mean())
print("Desviación estándar:",scores.std())

Accuracy promedio: 0.9733
Desviación estándar: 0.006165225056719346


Utiliza el AUC como métrico de Cross-Validation.

In [13]:
scores = cross_val_score(model, X_scaled, y, cv=kf, scoring='roc_auc')

print("AUC promedio:",scores.mean())
print("Desviación estándar: ",scores.std())

AUC promedio: 0.9490942119736451
Desviación estándar:  0.01705436005004184


Compara resultados.

Para el dataset Default, la regresión logística es el modelo más apto, con un accuracy promedio de 0.9731 y un rendimiento estable usando K-Folds, mostrando también un AUC alto (0.9491) que indica buena capacidad de separar clientes que pagan de los que no. Entre los SVM, el kernel RBF es el más preciso (accuracy 0.9716), seguido de Poly y Linear, mientras que Sigmoid funciona peor. En general, todos los modelos distinguen bastante bien las clases, pero la regresión logística combina precisión, estabilidad y simplicidad, haciéndola la opción más recomendable.


# Addendum

Métricos disponibles para clasificación:
- ‘accuracy’
- ‘balanced_accuracy’
- ‘top_k_accuracy’
- ‘average_precision’
- ‘neg_brier_score’
- ‘f1’
- ‘f1_micro’
- ‘f1_macro’
- ‘f1_weighted’
- ‘f1_samples’
- ‘neg_log_loss’
- ‘precision’ etc.
- ‘recall’ etc.
- ‘jaccard’ etc.
- ‘roc_auc’
- ‘roc_auc_ovr’
- ‘roc_auc_ovo’
- ‘roc_auc_ovr_weighted’
- ‘roc_auc_ovo_weighted’
- ‘d2_log_loss_score’

# References

[1] Shigeo Abe.Support Vector Machines for Pattern Classification,2Ed.Springer-Verlag London,2010. ISBN978-1-84996-097-7. URLhttps://www.springer.com/gp/book/9781849960977.

[2] Johan A K Suykens, Tony Van Gestel, Jos De Brabanter, BartDe Moor, and Joos Vandewalle.Least Squares Support VectorMachines. World Scientific,2002. ISBN9789812381514. URLhttps://www.worldscientific.com/worldscibooks/10.1142/5089.

[3] Bradley, A. P. (1997). The use of the area under the ROC curve in the evaluation of machine learning algorithms. Pattern recognition, 30(7), 1145-1159. URL https://www.researchgate.net/post/how_can_I_interpret_the_ROC_curve_result